pkg/instrumentation/flags.go

```diff
@@ -21,135 +21,209 @@ import (
 
 // Our instrumentation options.
 var opt = defaultOptions().(*options)
 
-// MarshalJSON is the JSON marshaller for TraceConfig values.
-func (tc TraceConfig) MarshalJSON() ([]byte, error) {
-	switch {
-	case tc <= 0.005:
-		return json.Marshal("disabled")
-	case tc <= 0.1:
-		return json.Marshal("production")
-	case tc >= 0.95:
-		return json.Marshal("testing")
-	}
-	return json.Marshal(tc)
+// MarshalJSON is the JSON marshaller for Sampling values.
+func (s Sampling) MarshalJSON() ([]byte, error) {
+	return json.Marshal(s.String())
 }
 
-// UnmarshalJSON is the JSON unmarshaller for TraceConfig values.
-func (tc *TraceConfig) UnmarshalJSON(raw []byte) error {
+// UnmarshalJSON is the JSON unmarshaller for Sampling values.
+func (s *Sampling) UnmarshalJSON(raw []byte) error {
 	var obj interface{}
 	if err := json.Unmarshal(raw, &obj); err != nil {
-		return instrumentationError("failed to unmarshal TraceConfig value:%v", err)
+		return instrumentationError("failed to unmarshal Sampling value: %v", err)
 	}
-	switch obj.(type) {
+	switch v := obj.(type) {
 	case string:
-		switch strings.ToLower(obj.(string)) {
-		case "disabled":
-			*tc = Disabled
-		case "testing":
-			*tc = Testing
-		case "production":
-			*tc = Production
-		default:
-			return instrumentationError("invalid TraceConfig value '%s'", obj.(string))
+		if err := s.Parse(v); err != nil {
+			return err
 		}
 	case float64:
-		*tc = obj.(TraceConfig)
+		*s = Sampling(v)
+	default:
+		return instrumentationError("invalid Sampling value of type %T: %v", obj, obj)
+	}
+	return nil
+}
+
+// Parse parses the given string to a Sampling value.
+func (s *Sampling) Parse(value string) error {
+	switch strings.ToLower(value) {
+	case "disabled":
+		*s = Disabled
+	case "testing":
+		*s = Testing
+	case "production":
+		*s = Production
 	default:
-		return instrumentationError("invalid TraceConfig value of type %T: %v", obj, obj)
+		f, err := strconv.ParseFloat(value, 64)
+		if err != nil {
+			return instrumentationError("invalid Sampling value '%s': %v", value, err)
+		}
+		*s = Sampling(f)
 	}
 	return nil
 }
 
-// Sampler returns a trace.Sampler corresponding to the TraceConfig value.
-func (tc TraceConfig) String() string {
+// String returns the Sampling value as a string.
+func (s Sampling) String() string {
 	switch {
-	case tc <= 0.005:
+	case s == Disabled:
 		return "disabled"
-	case tc <= 0.1:
+	case s == Production:
 		return "production"
-	case tc >= 0.95:
+	case s == Testing:
 		return "testing"
 	}
-	return strconv.FormatFloat(float64(tc), 'f', -1, 64)
+	return strconv.FormatFloat(float64(s), 'f', -1, 64)
 }
 
-// Sampler returns a trace.Sampler corresponding to the TraceConfig value.
-func (tc TraceConfig) Sampler() trace.Sampler {
+// Sampler returns a trace.Sampler corresponding to the Sampling value.
+func (s Sampling) Sampler() trace.Sampler {
 	switch {
-	case tc >= 0.95:
+	case s == Testing:
 		return trace.AlwaysSample()
-	case tc <= 0.005:
+	case s == Disabled:
```

Comment:
Hmm, I think we should keep the `<=` and `>=` operators here, right(?) And prolly re-define `Disabled = 0.005` or smth.

Reply:
Yes, and have symmetrical cutoffs at both ends (although, now that I look at it, I don't think they make much sense).